docs/spec/concepts.rst

```diff
@@ -3,97 +3,216 @@
+static type error. This allows very few operations! For example, if ``x`` is
+typed as ``object``, ``x.foo`` should be a static type error, because not all
+Python objects have an attribute ``foo``.
+
+A term typed as :ref:`Any`, on the other hand, should be assumed to have
+_some_ specific static type, but _which_ static type is not known. A static
+checker should not emit any errors that depend on assuming a particular static
+type; a static checker should instead assume that the runtime is responsible
+for checking the type of operations on this term, as in a dynamically-typed
+language.
+
+The gradual guarantee
+~~~~~~~~~~~~~~~~~~~~~
+
+:ref:`Any` allows gradually adding static types to a dynamically-typed program.
+For a fully dynamically-typed program, a static checker has the type :ref:`Any`
+for all terms, and should emit no errors. Adding more static type annotations
+to the program (making the program more statically typed) may result in static
+type errors, if the program is not correct or if the static type annotations
+aren't able to fully represent the runtime types. Removing static type
+annotations (making the program more dynamic) should not result in additional
+static type errors. This is often referred to as the **gradual guarantee**.
+
+In Python's type system, we don't take the gradual guarantee as a strict
+requirement, but it's a useful guideline.
+
+The subtype relation
+--------------------
+
+A static type ``B`` is a **subtype** of another static type ``A`` if and only
+if the set of values represented by ``B`` is a subset of the set of values
+represented by ``A``. Because the subset relation on sets is transitive and
+reflexive, the subtype relation is also transitive (if ``C`` is a subtype of
+``B`` and ``B`` is a subtype of ``A``, then ``C`` is a subtype of ``A``) and
+reflexive (``A`` is always a subtype of ``A``).
+
+The **supertype** relation is the inverse of subtype: ``A`` is a supertype of
+``B`` if and only if ``B`` is a subtype of ``A``; or equivalently, if and only
+if the set of values represented by ``A`` is a superset of the values
+represented by ``B``. The supertype relation is also transitive and reflexive.
+
+We also define an **equivalence** relation on static types: the types ``A`` and
+``B`` are equivalent (or "the same type") if and only if ``A`` is a subtype
+of ``B`` and ``B`` is a subtype of ``A``. This means that the set of values
+represented by ``A`` is both a superset and a subset of the values represented
+by ``B``, meaning ``A`` and ``B`` must represent the same set of values.
+
+We may describe a type ``B`` as "narrower" than a type ``A`` (or as a "strict
+subtype" of ``A``) if ``B`` is a subtype of ``A`` and ``B`` is not equivalent
+to ``A``.
+
+The consistency relation
+------------------------
+
+Since :ref:`Any` represents an unknown static type, it does not represent any
+known single set of values, and thus it is not in the domain of the subtype,
```

Comment:
This phrasing may not be the best and I'm not sure how to phrase this well, but I do think that the point is actually correct.

The type `list[object]` is *exactly* the set of all possible `list` objects (and is a fully static type that permits exactly the operations that are legal on this entire set).

The type `list[Any]`, on the other hand, represents a list of objects each of whose type is unknown, and so under gradual typing rules we permit each element to be used in arbitrary operations (by way of materializing as in the next paragraph).

Maybe actually discussing this example would clarify things a bit. Using examples may not feel very spec-y, but there's a reason why even math textbooks usually intermingle their abstract definitions with concrete examples.

Reply:
A concise way to think of `list[Any]` is that it is an arbitrary set of list objects, not necessarily all.

(And `list[object]` is subtle because `list` is an invariant generic.  It's the set of all lists that allow inserting `object`s.)

Reply:
I think the question of whether `list[object]` represents all possible lists is related to the other concern I had (not about this change but about typing spec and docs in general) that I want to be sure we don't use language implying a *value* has a type.

Any list value certainly can be a valid `list[object]`, but variance is a good example of why values don't actually have types, because what type operations are allowed is a question not only about the value but about all possible references to it. This is why static types in general only apply to terms, not to values.

So I guess here I was getting sloppy about my language in exactly the way I worry will leak into the spec and docs :)

Reply:
I would approach this last point slightly differently. I think for a model of "types as sets of values" to work, we need to include the type signatures of methods in our concept of what a "value" is. So a list value with a method `append(self, obj: int) -> None` is distinct from a list value with a method typed `append(self, obj: object) -> None`.

Reply:
Oh, I see what you were saying :)

You have to have a different notion of **value**.  A value is a python object tagged with a fully static type.  (And an unspecialized generic like `list` is not a type.)  So the set of lists tagged with `list[object]` and the set tagged with `list[int]` are really distinct sets of values.

If the type system inserted runtime checked casts, it would use this tag.  But CPython doesn't do that so we can have an implementation that erases the type tags.

Reply:
I added the example `tuple[int, Any]` above to try to help clarify this distinction. I think a deeper discussion of the generalization to invariant and contravariant generics probably doesn't belong here, but rather in the generics chapter.

Resolving this thread for now, feel free to reopen if you think there's more that needs to be addressed here.